### PR TITLE
Create adaptive_merge_sort.cpp

### PR DESCRIPTION
**Pull request by anant-jain01**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Adaptive merge sort code

### CODE REVIEW

**Comment by anant-jain01** [comment]
@Panquesito7 @realstealthninja please do the needful

File: sorting/adaptive_merge_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+ * @details This file contains the implementation of the Adaptive Merge Sort algorithm.
+ * The algorithm is a stable, divide-and-conquer based sorting algorithm.
+ *
+ * @see https://medium.com/@kaweendra/the-ultimate-sorting-algorithm-6513d6968420
+ */
+
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <cassert>
```
Comment (realstealthninja) [inline sorting/adaptive_merge_sort.cpp:14]: document headers please

File: sorting/adaptive_merge_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    return array;
+}
+
+/**
+ * @brief Unit tests for the AdaptiveMergeSort algorithm
+ */
+void testAdaptiveMergeSort() {
+    // Test Integer Input
+    std::vector<int> integers = {4, 23, 6, 78, 1, 54, 231, 9, 12};
+    std::vector<int> expectedIntegers = {1, 4, 6, 9, 12, 23, 54, 78, 231};
```
Comment (realstealthninja) [inline sorting/adaptive_merge_sort.cpp:81]: add test case for negative numbers

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by anant-jain01** [comment]
@realstealthninja @Panquesito7 please recheck and approve under hacktoberfest tag...

**Review by realstealthninja — CHANGES_REQUESTED** [review]
